cranelift/codegen/src/isa/x64/inst/external.rs

```diff
@@ -126,9 +128,59 @@ impl asm::AsReg for PairedGpr {
 
 /// A pair of XMM registers, one for reading and one for writing.
 #[derive(Clone, Copy, Debug)]
+#[expect(missing_docs, reason = "self-describing variants")]
 pub struct PairedXmm {
-    pub(crate) read: Xmm,
-    pub(crate) write: WritableXmm,
+    pub read: Xmm,
+    pub write: WritableXmm,
+}
+
+impl From<WritableXmm> for PairedXmm {
+    fn from(wxmm: WritableXmm) -> Self {
+        let read = wxmm.to_reg();
+        let write = wxmm;
+        Self { read, write }
+    }
+}
+
+/// For ABI ergonomics.
+impl From<WritableXmm> for asm::Xmm<PairedXmm> {
+    fn from(wgpr: WritableXmm) -> Self {
+        asm::Xmm::new(wgpr.into())
+    }
+}
+
+// For emission ergonomics.
+impl From<Writable<Reg>> for asm::Xmm<PairedXmm> {
+    fn from(wxmm: Writable<Reg>) -> Self {
+        assert!(wxmm.to_reg().class() == RegClass::Float);
+        let wxmm = WritableXmm::from_writable_reg(wxmm).unwrap();
+        Self::new(wxmm.into())
+    }
+}
+
+// For emission ergonomics.
+impl From<Reg> for asm::Xmm<Xmm> {
+    fn from(xmm: Reg) -> Self {
+        assert!(xmm.class() == RegClass::Float);
+        let xmm = Xmm::unwrap_new(xmm);
+        Self::new(xmm)
+    }
+}
+
+// For emission ergonomics.
+impl From<Reg> for asm::XmmMem<Xmm, Gpr> {
+    fn from(xmm: Reg) -> Self {
+        assert!(xmm.class() == RegClass::Float);
+        let xmm = Xmm::unwrap_new(xmm);
+        Self::Xmm(xmm)
+    }
+}
```

Comment:
I'm not particularly happy with how much conversion code is needed. Having this here is certainly better than constructing all the necessary structs when trying to reason about some code sequence, but it still feels a bit much. This is not really anything that needs to be solved in this PR; in fact, it might be better solved once most instructions have been ported to the new assembler. But I bring it up to see if anyone has any good ideas for reducing this kind of thing.

Reply:
Yeah, I suspect we'll be able to think a little more clearly about this (or at least I know I will!) once we're able to DCE away all of the types that the old-style insts are keeping alive...